Beta version cannot open existing database using passcode

- use app version v2018.02.26.981 to set up a new database
- open settings/security/acivate passcode
- update app to latest beta
- it will keep crashing until the database has been deleted from the device

I can reproduce this in android studio using the attached data.mmb (passcode: 1234)

[data.zip](https://github.com/moneymanagerex/android-money-manager-ex/files/2752920/data.zip)

Without passcode it opens, after activating the passcode:
![fatal](https://user-images.githubusercontent.com/1249745/51083245-01006300-1717-11e9-93b3-577bb28ce461.png)
```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.money.manager.ex.debug, PID: 3376
    java.lang.NoClassDefFoundError: Failed resolution of: Landroid/hardware/fingerprint/FingerprintManager;
        at com.money.manager.ex.PasscodeActivity.onCreate(PasscodeActivity.java:164)
        at android.app.Activity.performCreate(Activity.java:5990)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1106)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2278)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2387)
        at android.app.ActivityThread.access$800(ActivityThread.java:151)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1303)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:135)
        at android.app.ActivityThread.main(ActivityThread.java:5254)
        at java.lang.reflect.Method.invoke(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:372)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "android.hardware.fingerprint.FingerprintManager" on path: DexPathList[[zip file "/data/app/com.money.manager.ex.debug-2/base.apk", zip file "/data/app/com.money.manager.ex.debug-2/split_lib_dependencies_apk.apk", zip file "/data/app/com.money.manager.ex.debug-2/split_lib_slice_0_apk.apk", zip file "/data/app/com.money.manager.ex.debug-2/split_lib_slice_1_apk.apk", zip file "/data/app/com.money.manager.ex.debug-2/split_lib_slice_2_apk.apk", zip file "/data/app/com.money.manager.ex.debug-2/split_lib_slice_3_apk.apk", zip file "/data/app/com.money.manager.ex.debug-2/split_lib_slice_4_apk.apk", zip file "/data/app/com.money.manager.ex.debug-2/split_lib_slice_5_apk.apk", zip file "/data/app/com.money.manager.ex.debug-2/split_lib_slice_6_apk.apk", zip file "/data/app/com.money.manager.ex.debug-2/split_lib_slice_7_apk.apk", zip file "/data/app/com.money.manager.ex.debug-2/split_lib_slice_8_apk.apk", zip file "/data/app/com.money.manager.ex.debug-2/split_lib_slice_9_apk.apk"],nativeLibraryDirectories=[/vendor/lib, /system/lib]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
        at com.money.manager.ex.PasscodeActivity.onCreate(PasscodeActivity.java:164) 
        at android.app.Activity.performCreate(Activity.java:5990) 
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1106) 
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2278) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2387) 
        at android.app.ActivityThread.access$800(ActivityThread.java:151) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1303) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:135) 
        at android.app.ActivityThread.main(ActivityThread.java:5254) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at java.lang.reflect.Method.invoke(Method.java:372) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:903) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:698) 
    	Suppressed: java.lang.ClassNotFoundException: android.hardware.fingerprint.FingerprintManager
        at java.lang.Class.classForName(Native Method)
        at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
        at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
        		... 15 more
     Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack available
E/SQLiteLog: (284) automatic index on sqlite_sq_A3807FB0(id)
E/SQLiteLog: (284) automatic index on sqlite_sq_A3807E70(id)
E/SQLiteLog: (284) automatic index on sqlite_sq_A5C08900(accountid)
```

--- a/app/src/main/java/com/money/manager/ex/PasscodeActivity.java
+++ b/app/src/main/java/com/money/manager/ex/PasscodeActivity.java
@@ -20,7 +20,7 @@ import android.Manifest;
 import android.app.KeyguardManager;
 import android.content.Intent;
 import android.content.pm.PackageManager;
-import android.hardware.fingerprint.FingerprintManager;
+import androidx.core.hardware.fingerprint.FingerprintManagerCompat;
 import android.os.Build;
 import android.os.Bundle;
 import androidx.core.app.ActivityCompat;
@@ -75,8 +75,8 @@ public class PasscodeActivity extends AppCompatActivity {
 	private Cipher cipher;
 	private KeyStore keyStore;
 	private KeyGenerator keyGenerator;
-	private FingerprintManager.CryptoObject cryptoObject;
-	private FingerprintManager fingerprintManager;
+	private FingerprintManagerCompat.CryptoObject cryptoObject;
+	private FingerprintManagerCompat fingerprintManager;
 	private KeyguardManager keyguardManager;
 
 	@Override
@@ -161,7 +161,8 @@ public class PasscodeActivity extends AppCompatActivity {
 		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
 
 			keyguardManager = (KeyguardManager) getSystemService(KEYGUARD_SERVICE);
-			fingerprintManager = (FingerprintManager) getSystemService(FINGERPRINT_SERVICE);
+			//fingerprintManager = (FingerprintManager) getSystemService(FINGERPRINT_SERVICE);
+			fingerprintManager = FingerprintManagerCompat.from(getApplicationContext());
 
 			if (!fingerprintManager.isHardwareDetected()) {
 				((ImageView) findViewById(R.id.fpImageView))
@@ -189,7 +190,7 @@ public class PasscodeActivity extends AppCompatActivity {
 						e.printStackTrace();
 					}
 					if (initCipher()) {
-						cryptoObject = new FingerprintManager.CryptoObject(cipher);
+						cryptoObject = new FingerprintManagerCompat.CryptoObject(cipher);
 						FingerprintHandler helper = new FingerprintHandler(this);
 						helper.startAuth(fingerprintManager, cryptoObject);
 					}

--- a/app/src/main/java/com/money/manager/ex/home/MainActivity.java
+++ b/app/src/main/java/com/money/manager/ex/home/MainActivity.java
@@ -295,7 +295,8 @@ public class MainActivity
     protected void onActivityResult(int requestCode, int resultCode, Intent data) {
         super.onActivityResult(requestCode, resultCode, data);
 
-        if (resultCode != RESULT_OK) return;
+        // don't accidentally bypass passcode (failures), e.g. pressing physical back button
+        if (resultCode != RESULT_OK && requestCode != RequestCodes.PASSCODE) return;
 
         switch (requestCode) {
 //            case RequestCodes.SELECT_FILE:

--- a/app/src/main/java/com/money/manager/ex/passcode/FingerprintHandler.java
+++ b/app/src/main/java/com/money/manager/ex/passcode/FingerprintHandler.java
@@ -8,14 +8,14 @@ import android.content.Context;
 import android.content.Intent;
 import android.Manifest;
 import android.content.pm.PackageManager;
-import android.hardware.fingerprint.FingerprintManager;
-import android.os.CancellationSignal;
+import androidx.core.hardware.fingerprint.FingerprintManagerCompat;
+import androidx.core.os.CancellationSignal;
 import androidx.core.app.ActivityCompat;
 import androidx.appcompat.app.AppCompatActivity;
 import android.widget.Toast;
 import com.money.manager.ex.PasscodeActivity;
 
-public class FingerprintHandler extends FingerprintManager.AuthenticationCallback {
+public class FingerprintHandler extends FingerprintManagerCompat.AuthenticationCallback {
 
     private CancellationSignal cancellationSignal;
     private Context context;
@@ -24,12 +24,12 @@ public class FingerprintHandler extends FingerprintManager.AuthenticationCallbac
         context = mContext;
     }
 
-    public void startAuth(FingerprintManager manager, FingerprintManager.CryptoObject cryptoObject) {
+    public void startAuth(FingerprintManagerCompat manager, FingerprintManagerCompat.CryptoObject cryptoObject) {
         cancellationSignal = new CancellationSignal();
         if (ActivityCompat.checkSelfPermission(context, Manifest.permission.USE_FINGERPRINT) != PackageManager.PERMISSION_GRANTED) {
             return;
         }
-        manager.authenticate(cryptoObject, cancellationSignal, 0, this, null);
+        manager.authenticate(cryptoObject, 0, cancellationSignal, this, null);
     }
 
     @Override
@@ -48,7 +48,7 @@ public class FingerprintHandler extends FingerprintManager.AuthenticationCallbac
     }
 
     @Override
-    public void onAuthenticationSucceeded(FingerprintManager.AuthenticationResult result) {
+    public void onAuthenticationSucceeded(FingerprintManagerCompat.AuthenticationResult result) {
         //Toast.makeText(context, "Authentication Success!", Toast.LENGTH_LONG).show();
 
         Intent data = new Intent();

--- a/app/src/main/res/layout-land/passcode_activity.xml
+++ b/app/src/main/res/layout-land/passcode_activity.xml
@@ -118,6 +118,23 @@
                     android:maxLines="1" />
             </LinearLayout>
         </LinearLayout>
+
+        <ImageView
+            android:layout_width="75dp"
+            android:layout_height="75dp"
+            app:srcCompat="@drawable/fingerprint"
+            android:id="@+id/fpImageView"
+            android:layout_below="@+id/textView"
+            android:layout_centerHorizontal="true"
+            android:layout_marginTop="36dp"/>
+
+        <com.money.manager.ex.view.RobotoTextView
+            android:id="@+id/fingerprintInfo"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:textSize="@dimen/mmx_text_view_size_small"
+            android:text="@string/fingerprint_info"/>
+
     </LinearLayout>
 
     <LinearLayout

